application/src/main/kotlin/net/corda/v5/application/crypto/SigningService.kt

```diff
@@ -20,10 +21,36 @@ import java.util.*
 @DoNotImplement
 interface SigningService {
 
+    /**
+     * Using the provided signing [PublicKey], internally chooses the most appropriate signing spec,
+     * looks up the matching [PrivateKey] and signs the data, selecting the most appropriate signing type.
+     * This overload is preferred; see the other overload if you need to control the signing type.
+     *
+     * @param payload The data to sign over using the chosen key.
+     * @param metadata Optional map of string configuration metadata to be included in the signature. If the metadata
+     *                 is null, or if there are no know entries in the map, that's equivalent to calling `sign` without
+     *                 metadata.
+     * @param publicKey The [PublicKey] partner to an internally held [PrivateKey], either derived from the node's
+     * primary identity, or previously generated via the freshKey method. If the [PublicKey] is actually
+     * a [CompositeKey], the first leaf signing key hosted by the node is used.
+     * @param digestAlgorithmName the message digest algorithm name; if null have the crypto service choose one.
+     * @return A [DigitalSignature.WithKey] representing the signed data and the [PublicKey] that belongs to the
+     * same [KeyPair] as the [PrivateKey] that signed the data. The signature type will be determined automatically.
+     *
+     * @throws [CordaRuntimeException] If the input key is not a member of [keys].
+     */
+    @Suspendable
+    fun sign(payload: ByteArray, metadata: Map<String, String>?, publicKey: PublicKey, digestAlgorithmName: DigestAlgorithmName?): DigitalSignature.WithKey
```

Comment:
The return values should be changed to `DigitalSignature.WithMetadata` where it's defined as

```kotlin
/**
 * A wrapper around a digital signature.
 *
 * @property bytes The signature.
 * @param bytes The signature.
 */
@CordaSerializable
open class DigitalSignature(
    bytes: ByteArray
) : OpaqueBytes(bytes) {
    /**
     * A digital signature that identifies who the public key is owned by.
     *
     * @param by The public key of the corresponding private key used to sign the data (as if an instance
     * of the [CompositeKey] is passed to the sign operation it may contain keys which are not actually owned by
     * the member).
     * @param bytes The signature.
     * @param context The context which was passed to the signing operation, note that this context is not signed over.
     */
    open class WithKey(
        /**
         * Public key which corresponding private key was used to sign the data (as if an instance
         * of the [CompositeKey] is passed to the sign operation it may contain keys which are not actually owned by
         * the member).
         */
        val by: PublicKey,
        bytes: ByteArray,
        /**
         * The context which was passed to the signing operation, note that this context is not signed over.
         */
        val context: Map<String, String>
    ) : DigitalSignature(bytes)

    class WithMetadata(
        /**
         * Public key which corresponding private key was used to sign the data (as if an instance
         * of the [CompositeKey] is passed to the sign operation it may contain keys which are not actually owned by
         * the member).
         */
        by: PublicKey,
        bytes: ByteArray,
        /**
         * The context which was passed to the signing operation, note that this context is not signed over.
         */
        context: Map<String, String>,
        /**
         * The serialised metadata which was used by the signing operation, note that the metadata **is signed over**.
         * The actual signing data is concatenation of the serialised data and the user data.
         */
        metadata: ByteArray,
    ) : WithKey(by, bytes, context)
}
```

Reply:
Done